Refactor tests which use debug_dtm_action to use faultinjector

vbwagner opened the pull request:
Refactor tests which use debug_dtm_action to use faultinjector

gpdb tests use two mechanisms to generate artificial faults.

1. gp_fault_injector extension with corresponding backend code;
2. set of GUC variables with names beginning with `debug_dtm_action`.

This patch changes all tests from `debug_dtm_action` to faultinjector.

It includes following extensions to faultinjector:

1. Additional condition to choose whether to throw fault or not —
   subtransaction nesting level;
2. Compatibility SQL function which allows avoiding changes in old test which
   use `gp_inject_fault` without that new condition;
3. Extending `DDLStatement_e` enumeration to include all SQL operators and DTX
   protocol messages used in tests, which previously used `debug_dtm_action`;
4. wrappers to fault injection point function which translate textual command
   tags and DTX protocol message code to `DDLStatement_e`;
5. Improved README for `gp_inject_fault` extension, which documents these
   changes and also documents some already existing functionality which was
   left undocumented.

Apart from faultinjector changes patch includes following changes:

1. Replacing of code blocks which check `debug_dtm_action` variables with
   calls to faultinjector macros;
2. Removal of `debug_dtm_action_*` GUC variables from guc_gp.cl
3. Replacing of `SET debug_dtm_action*` in the tests to calls to
   `gp_inject_fault` function;
4. Modifications of expected test results which include changed text of
   injected error and echo of SQL statements which inject an error.

vbwagner committed:
Refactoring of fault_injector to support former debug_dtm_action points

Following changes are made to faultinjector code

1. In the gp_inject_fault extension, one extra parameter added to C
   language function gp_inject_fault - nestingLevel for nesting
   subtransaction
2. Compatibility SQL function with same number of argumetns as former C
   function is created to avoid incompatibilites.
3. DDLStatement_e enumeration is extended to cover all SQL and DTX
   statements which are used in tests with debug_dtm_action (they are
   not all ddl, but...)
4. Since string names of some of these statements as they are used in
   the code, contain spaces, scanf format in HandleFaultMessage is
   changed to handle strings with spaces.
5. Parameter nestingLevel is added to function
   FaultInjector_InjectFaultIfSet_out_of line.
6. Since SQL Statement Tag present in the context where fault injection
   code is called as string, created new function
   FaultInjector_InjectFaultIfSet_SQL_out_of_line and corresponding
   macro which wraps basic fault injection function and  translates
   string statement info DDLStatement_e value
7. Since DTXProtocol command is available as enumeration, but not same
   as DLLStatement_e, recoding table is added to faultinjector.c and
   function FaultInjector_InjectFaultIfSet_DTX_out of_line, which
   analogous to prevoius step.
8. To make definition of DtxtProtocolCommand enumeration available  in
   the context of faultinjector.h (where function from previous point
   is declared) header files postgres.h and cdb/cdbtm.h are included
   into this file.

vbwagner committed:
Replace debug_dtm_action fault injection points with faultinjector calls

There were three functions where debug_dtm_action GUC variables were
checked and faults thrown:

exec_mpp_query
exec_simple_query and
exec_mpp_dtx_protocol_command

First two have two ereport calls at start and at end of functions.
Last have three ereport calls because at start it was possible to throw
either ERROR or PANIC. Faultinjector is able to throw both error and
panic from any incection point, so there are now just two
injection points in all three functions.

Functions exec_mpp_query and exec_simple_query have string variable
commandTag with printable of command. So we use this value to define
ddlStatement condition

For exec_mpp_dtx_query we use dtxProtocolCommand with enumerated type
DtxProtocolCommand.

Injection points are named

dtm_exec_mpp_query_start
dtm_exec_mpp_query_end
dtm_exec_simple_command_start
dtm_exec_simple_command_end
exec_mpp_dtx_protocol_start
exec_mpp_dtx_protocol_end

Old static helper functions CheckDebugDtmActionSqlCommandTag and
CheckDebugDtmActionProtocol are removed.

vbwagner committed:
Refactor tests which use debug_dtm_action

This commit replaces all references for debug_dtm_action* GUC variables
in the tests (including helper C program twophase_pqexecparams.c)
with calls to gp_inject_fault function.

This also requires changes to src/test/regress/greenplum_schedule
because there were no check that tests which inject faults using
debug_dtm_action GUC variables are not executed as part of parallel
group, but such check present for tests which use gp_inject_fault.

And it is probaly right thing because injected fault may unexpectedly
disrupt parallel sessions.

If GUC variables are very seldom reset before next test, gp_inject_fault
requires reset each time before injecting to same injection point with
different parameters.

vbwagner committed:
Remove debug_dtm_action* GUC variables

This commit removes all debug_dtm_action* GUC variables and corresponing
global variables, defines, string-to-enum recoding tables.

It is made separate commit because we have to be sure that our
replacement of these variables with gp_inject_fault function is
complete.

vbwagner committed:
Expected results for test wihich use debug_dtm_actions

Expected results of these tests have to be altered, because
expeced results contain all SQL commands and their output,
so SET debug_dtm_action* have to be removed and
SELECT gp_inject_fault(...) FROM... added.

Also, when injected fault message not caught, it also have to be
changed.

Commit:
Improved README for gp_inject_fault extension

Documented additional fault injection macros and parameters
required to replace debug_dtm_action variables, and documented
some previously undocumented parts of faultinjector.

Commit:
Add symbols, changed by the patch to .abi-check/ignore

Refactoring of faultinjector and removing GUC variables cause
ABI changes. This change is harmless because this GUC variables
are never used outside tests and faultinjector is not compiled in
in production builds.  Moreover, compatibility macros are
provided to preserve to FaultInjector_InjectFaultIfSet behavoir

Commit:
Fix list of changed symbols for abi-checker

RekGRpth commented:
> Removal of `debug_dtm_action_*` GUC variables form guc_gp.c

Could anyone in the release build use these GUCs?

RekGRpth commented on gpcontrib/gp_inject_fault/gp_inject_fault.c:
```diff
@@ -142,16 +144,17 @@ gp_inject_fault(PG_FUNCTION_ARGS)
 			tableName = "#";
-		snprintf(msg, 1024, "faultname=%s type=%s ddl=%s db=%s table=%s "
-				 "start=%d end=%d extra=%d sid=%d ",
+		snprintf(msg, 1024, "faultname=%s type=%s ddl='%s' db=%s table=%s "
```
Why was this change made?

vbwagner replied:
Because we extend set of values for ddlStatement parameter. And now these values include such as MPPEXEC UTILITY. So we need such format of message, which allows to use spaces in the value.

RekGRpth commented on gpcontrib/gp_inject_fault/README:
```diff
@@ -133,16 +150,98 @@ status
+   you don't want limit injection by particular operator. See
+   `faultinjector_lists.h` for list of allowed operators. This
+   parameter makes sense only if injection point in the code was defined
+   by macro which accepts `ddlStatement` parameter and sime meaningful
```
maybe
```suggestion
   by macro which accepts `ddlStatement` parameter and some meaningful
```
?

vbwagner replied:
Fixed

RekGRpth commented on gpcontrib/gp_inject_fault/README:
```diff
@@ -133,16 +150,98 @@ status
+   by macro which accepts `ddlStatement` parameter and sime meaningful
+   value is passed to it.
+* database text - database name which should be processed. Makes sense
+    only if you know that database name was passed to injection point in
+	the code
```
```suggestion
   only if you know that database name was passed to injection point in
   the code
```

vbwagner replied:
fixed

RekGRpth commented on gpcontrib/gp_inject_fault/README:
```diff
@@ -133,16 +150,98 @@ status
+	the code
+*  tablename text - table name which should be processed. Same
+   conditions as in previous case stands.
+* start_occurrence int4 - number of times injection point is hit when
+   first fault is injected.
+*  end_occurrence int4 - number of times injection point is hit when
+   last fault is injected. Use -1 if you don't want to limit number of
```
Somehow the formatting is different, somewhere after the asterisk there is one space, and somewhere two, you decide.

vbwagner replied:
fixed

RekGRpth commented on gpcontrib/gp_inject_fault/README:
```diff
@@ -133,16 +150,98 @@ status
+Fault Injection Macros
+----------------------
+
+In the C code fault injection point is marked by adding one on macros,
```
maybe
```suggestion
In the C code fault injection point is marked by adding one of macros,
```
?

vbwagner replied:
Fixed

RekGRpth commented on src/backend/tcop/postgres.c:
```diff
@@ -1351,16 +1348,7 @@ exec_mpp_query(const char *query_string,
-							Debug_dtm_action, commandTag)));
-		}
-
+		FaultInjector_InjectFaultIfSet_SQL("dtm_exec_mpp_query_start",commandTag,0);
```
pgindent your changes here and below

vbwagner replied:
Fixed

RekGRpth commented on src/backend/utils/misc/faultinjector.c:
```diff
@@ -2,7 +2,7 @@
  * faultinjector.c
  *	  GP Fault Injectors are used for Greenplum internal testing only.
- * 
+ *
```
I think it's worth removing formatting changes here and below that aren't relevant to this patch.

whitehawk commented on gpcontrib/gp_inject_fault/gp_inject_fault--1.0.sql:
```diff
@@ -14,11 +14,28 @@ CREATE FUNCTION @extschema@.gp_inject_fault(
 AS 'MODULE_PATHNAME'
 LANGUAGE C VOLATILE STRICT NO SQL;
 
+-- simpler versiion without nesting level
```
```suggestion
-- simpler version without nesting level
```

whitehawk commented:
It looks we need to do some cleanup in [guc.h](https://github.com/arenadata/gpdb/blob/7f1029f2a218e25d3946206604222c90b1f3cf82/src/include/utils/guc.h#L381) and [unsync_guc_name.h](https://github.com/arenadata/gpdb/blob/7f1029f2a218e25d3946206604222c90b1f3cf82/src/include/utils/unsync_guc_name.h#L64)

whitehawk commented on src/test/isolation2/sql/udf_exception_blocks_panic_scenarios.sql:
```diff
@@ -1,34 +1,10 @@
 -- Tests exception handling of GPDB PL/PgSQL UDF
 -- It exercises:
---  1. PROTOCOL or SQL type of dtm_action_target
+--  1. Several transaction control SQL stateement
```
```suggestion
--  1. Several transaction control SQL statement
```

whitehawk commented on src/test/isolation2/sql/udf_exception_blocks_panic_scenarios.sql:
```diff
@@ -1,34 +1,10 @@
--- which specific depth level in transaction to take the specified
--- dtm_action. This apples only to target with protocol and not SQL.
+--  3. dtx protocol command: subtransaction_begin, subtransaction_rollback or subtransaction_release
+--  4. Errors are teseted at beginning and end of command, panic only on
```
```suggestion
--  4. Errors are tested at beginning and end of command, panic only on
```

whitehawk commented:
In the PR description:
'to generate artifitial faults' -> 'to generate artificial faults'
'Removal of debug_dtm_action_* GUC variables form guc_gp.c' -> 'Removal of debug_dtm_action_* GUC variables from guc_gp.c'
?

whitehawk commented on src/test/regress/greenplum_schedule:
why do we need changes in the schedule file?

vbwagner replied:
> why do we need changes in the schedule file?

Because `make installcheck-world` doesn't allow tests, which use `gp_inject_fault` function to be executed as part of parallel test group. See commit message for commit [4b2c87649538f7d18](https://github.com/arenadata/gpdb/pull/1967/commits/4b2c87649538f7d18d2485c733023816c41c32b7).

whitehawk commented on src/test/regress/sql/udf_exception_blocks.sql:
If this test is executed after `distributed_transactions`, it fails:
```
============== running regression test queries        ==============
test distributed_transactions     ... ok         5415 ms (diff  174 ms)
test udf_exception_blocks         ... FAILED     5291 ms (diff  461 ms)
```

with 
```diff
diff -I HINT: -I CONTEXT: -I GP_IGNORE: -U3 /usr/re/share/review/ws115-ADBDEV-6161/gpdb/src/test/regress/expected/udf_exception_blocks.out /usr/re/share/review/ws115-ADBDEV-6161/gpdb/src/test/regress/results/udf_exception_blocks.out
--- /usr/re/share/review/ws115-ADBDEV-6161/gpdb/src/test/regress/expected/udf_exception_blocks.out      2025-08-25 17:39:00.141729694 +1000
+++ /usr/re/share/review/ws115-ADBDEV-6161/gpdb/src/test/regress/results/udf_exception_blocks.out       2025-08-25 17:39:00.192729696 +1000
@@ -744,20 +744,21 @@
 SELECT gp_inject_fault('exec_mpp_dtx_protocol_command_start','error',
                       'subtransaction_release', '','',1,-1,0,dbid)
   FROM gp_segment_configuration WHERE mode='s' and content=0 and role='p' ;
- gp_inject_fault
+ERROR:  Failure: could not insert fault injection, entry already exists (SOMEFILE:SOMEFUNC)
+DROP TABLE IF EXISTS employees;
+select test_protocol_allseg(1, 2,'f');
+ test_protocol_allseg
 ------------
- Success:
+
 (1 row)

-DROP TABLE IF EXISTS employees;
-select test_protocol_allseg(1, 2,'f');
-ERROR:  transaction  at level 1 already processed (SOMEFILE:SOMEFUNC)
-NOTICE:  catching the exception ...2
-CONTEXT:  PL/pgSQL function test_protocol_allseg(integer,integer,character) line 9 during exception cleanup
 select * from employees;
-ERROR:  relation "employees" does not exist
-LINE 1: select * from employees;
-                      ^
+ id | shop_id | gender
+----+---------+--------
+  0 |       1 | m
+  1 |       2 | f
+(2 rows)
+
 SELECT gp_inject_fault('exec_mpp_dtx_protocol_command_start','reset',dbid)
   FROM gp_segment_configuration WHERE mode='s' and content=0 and role='p' ;
  gp_inject_fault

```

I guess `distributed_transactions` doesn't reset some faults.

vbwagner replied:
This is definitely a problem which should be fixed.

vbwagner replied:
It seems that fault, which is not cleared is instered from C program twophase_pqexecparams.c

whitehawk commented on src/test/isolation2/sql/udf_exception_blocks_panic_scenarios.sql:
```diff
@@ -96,81 +72,82 @@ END; /* in func */
-SET debug_dtm_action_nestinglevel=0;
+--
+--
+SELECT gp_inject_fault('exec_mpp_dtx_protocol_command_start','panic','subtransaction_begin','','',1,-1,0,dbid)
```
Please add white spaces after commas here and elsewhere in the delta to improve readability, like below:
```suggestion
SELECT gp_inject_fault('exec_mpp_dtx_protocol_command_start', 'panic', 'subtransaction_begin', '', '', 1, -1, 0, dbid)
```

whitehawk commented on src/test/regress/expected/distributed_transactions.out:
```diff
@@ -91,34 +98,41 @@ INSERT INTO distxact1_3 VALUES (35);
 COMMIT;
 WARNING:  the distributed transaction 'Commit Prepared' broadcast failed to one or more segments. Retrying ... try 1
-DETAIL:  gid=66802167, state=Retry Commit Prepared
+DETAIL:  gid=16281, state=Retry Commit Prepared
```
I believe that testing infrastructure can handle the variable part of test output, like different output if no sorting is specified. For ex. I'm getting this test passed if I revert some part of the non-relevant delta:
```diff
diff --git a/src/test/regress/expected/distributed_transactions.out b/src/test/regress/expected/distributed_transactions.out
index 6deb6951700..35a12440025 100755
--- a/src/test/regress/expected/distributed_transactions.out
+++ b/src/test/regress/expected/distributed_transactions.out
@@ -108,19 +108,19 @@ SELECT gp_inject_fault('exec_mpp_dtx_protocol_command_start','error',

 COMMIT;
 WARNING:  the distributed transaction 'Commit Prepared' broadcast failed to one or more segments. Retrying ... try 1
-DETAIL:  gid=16281, state=Retry Commit Prepared
+DETAIL:  gid=66802167, state=Retry Commit Prepared
 NOTICE:  Releasing segworker group to retry broadcast.
 SELECT * FROM distxact1_3;
  a
 ----
- 34
- 37
+ 38
  32
- 33
  31
+ 33
+ 34
  35
  36
- 38
+ 37
 (8 rows)

 SELECT gp_inject_fault('exec_mpp_dtx_protocol_command_start','reset',dbid)
```  

Please re-check what parts of delta in the output files can be safely reverted in order to keep only relevant delta in the final patch.

vbwagner replied:
Probably adding an ORDER BY clause to SELECT statement would make test less flacky.

vbwagner replied:
> I believe that testing infrastructure can handle the variable part of test output, like different output if no sorting is specified. For ex. I'm getting this test passed if I revert some part of the non-relevant delta:

> 
>  SELECT gp_inject_fault('exec_mpp_dtx_protocol_command_start','reset',dbid)
> ```
> 
> Please re-check what parts of delta in the output files can be safely reverted in order to keep only relevant delta in the final patch.

Really diff tool used in the test is able to handle such differences. But if there is some other problems, diffs in the results of this select are included in test output too.

whitehawk replied:
Let's remove this unrelevant delta. Anyway on each test execution it will be different.

vbwagner replied:
> Let's remove this unrelevant delta. Anyway on each test execution it will be different.

Removal this delta means manual editiong of expected test results.  Typical way to handle changes in the test result is copying of *.out file from results to expected, after checking that everything found in regression.diffs is expected changes.

vbwagner replied:
I've  minimised diffs in expected results with base branch.

Commit:
Remove mention of debug_dtm_action vars

Remove mention of debug_dtm_action variables from comments in
udf_exception_blocks.sql, commented out SET statements from
twophase_pqexecparams.c and names from unsync_guc_name.h

Commit:
Fix formatting problems found during review

Change FaultInjector_InjectFault* calls in postgres.c to conform pgindent rules
Remove unneccessary whitespace changes from faultinjector.c

Commit:
Reset fault inserted by twophase_pqexecparams

C-language test program twophase_pqexecparams injects fault.
This fault is not resetted within program, so next test script would fail
if it tries to inject same fault before server is restarted.

We add resetting of this fault into calling SQL script, because we want to be sure that fault is resetted regradless of outcome of C program

RekGRpth commented on src/backend/utils/misc/faultinjector.c:
```diff
@@ -281,12 +281,66 @@ checkBgProcessSkipFault(const char* faultName)
+	return FaultInjector_InjectFaultIfSet_out_of_line(faultName, ddlStatement, "", "",
+											   nestingLevel);
+
+}
```
pgindent your changes

vbwagner replied:
Done

RekGRpth commented on src/include/utils/faultinjector_lists.h:
```diff
@@ -65,6 +65,29 @@ FI_DDL_STATEMENT(CreateTablespaces, "create_tablespaces")
 FI_DDL_STATEMENT(Truncate, "truncate")
 FI_DDL_STATEMENT(Vacuum, "vacuum")
+/* SQL statements */
+FI_DDL_STATEMENT(MPPExecUtility, "MPPEXEC UTILITY")
+FI_DDL_STATEMENT(MPPExecSelect, "MPPEXEC SELECT")
+FI_DDL_STATEMENT(MPPExecInsert, "MPPEXEC INSERT")
+FI_DDL_STATEMENT(MPPExecUpdate, "MPPEXEC UPDATE")
+FI_DDL_STATEMENT(MPPExecDelete, "MPPEXEC DELETE")
+FI_DDL_STATEMENT(Release,"RELEASE")
+FI_DDL_STATEMENT(Rollback,"ROLLBACK")
+FI_DDL_STATEMENT(SavePoint,"SAVEPOINT")
```
why not
```suggestion
FI_DDL_STATEMENT(MPPExecUtility, "mppexec_utility")
FI_DDL_STATEMENT(MPPExecSelect, "mppexec_select")
FI_DDL_STATEMENT(MPPExecInsert, "mppexec_insert")
FI_DDL_STATEMENT(MPPExecUpdate, "mppexec_update")
FI_DDL_STATEMENT(MPPExecDelete, "mppexec_delete")
FI_DDL_STATEMENT(Release, "release")
FI_DDL_STATEMENT(Rollback, "rollback")
FI_DDL_STATEMENT(SavePoint, "savepoint")
```
?

vbwagner replied:
Because these strings are commandTags, which are used not only  for fault injections. 
Tags for SQL statements are created by function CreateCommandTag declared in src/include/tcop/utility.h and used in 7 source files more than 30 times.

Command tags for MPPEXEC * statements are assigned inside exec_mpp_query function, but passed in two other functions besides fault injector. 

So we do not inveted our own string names for operators here, we use names already used in the code for other purposes.

Commit:
Fix typos in README file and test comments

whitehawk commented on src/backend/utils/misc/faultinjector.c:
```diff
@@ -358,7 +412,9 @@ FaultInjector_InjectFaultIfSet_out_of_line(
-
+		if (entryShared->nestingLevel !=0 && entryShared->nestingLevel != nestingLevel)
+		    /* fault injection is set for another nesting level */
+			break;
```
Let's add empty lines before and after the check, as it is already done for the surrounding existing checks.
```suggestion

		if (entryShared->nestingLevel !=0 && entryShared->nestingLevel != nestingLevel)
		    /* fault injection is set for another nesting level */
			break;
			
```

vbwagner replied:
Fixed

whitehawk commented on src/backend/utils/misc/faultinjector.c:
```diff
@@ -358,7 +412,9 @@ FaultInjector_InjectFaultIfSet_out_of_line(
 			/* fault injection is not set for the specified table name */
 			break;
-
+		if (entryShared->nestingLevel !=0 && entryShared->nestingLevel != nestingLevel)
```
```suggestion
		if (entryShared->nestingLevel != 0 && entryShared->nestingLevel != nestingLevel)
```

vbwagner replied:
Done by pgindent.

whitehawk commented on src/include/utils/faultinjector.h:
```diff
@@ -99,19 +102,46 @@ extern FaultInjectorType_e FaultInjector_InjectFaultIfSet_out_of_line(
 extern char *InjectFault(
 	char *faultName, char *type, char *ddlStatement, char *databaseName,
-	char *tableName, int startOccurrence, int endOccurrence, int extraArg, int gpSessionid);
+	char *tableName, int startOccurrence, int endOccurrence, int extraArg, int gpSessionid,int nestingLevel);
```
```suggestion
	char *tableName, int startOccurrence, int endOccurrence, int extraArg, int gpSessionid, int nestingLevel);
```

vbwagner replied:
fixed

whitehawk commented on src/include/utils/faultinjector.h:
```diff
@@ -99,19 +102,46 @@ extern FaultInjectorType_e FaultInjector_InjectFaultIfSet_out_of_line(
 #define FaultInjector_InjectFaultIfSet(faultName, ddlStatement, databaseName, tableName) \
 	(((*numActiveFaults_ptr) > 0) ? \
-	 FaultInjector_InjectFaultIfSet_out_of_line(faultName, ddlStatement, databaseName, tableName) : \
+	 FaultInjector_InjectFaultIfSet_out_of_line(faultName, ddlStatement, databaseName, tableName,0) : \
```
```suggestion
	 FaultInjector_InjectFaultIfSet_out_of_line(faultName, ddlStatement, databaseName, tableName, 0) : \
```

vbwagner replied:
Fixed

whitehawk commented on src/include/utils/faultinjector.h:
```diff
@@ -99,19 +102,46 @@ extern FaultInjectorType_e FaultInjector_InjectFaultIfSet_out_of_line(
+	const char*		statement,
+	int 			nestingLevel);
+
+/* Version of fault Inlectio funticion whict receives DTXProtocolCommand
```
this comment has lots of typos

vbwagner replied:
Fixed

silent-observer commented on src/test/regress/sql/distributed_transactions.sql:
```diff
@@ -58,14 +57,12 @@ INSERT INTO distxact1_2 VALUES (25);
-SET debug_dtm_action_protocol = "prepare";
+SELECT gp_inject_fault('exec_mpp_dtx_protocol_command_start','error','prepare',
+	'','',1,1,0,dbid)
+ FROM gp_segment_configuration WHERE mode='s' and content=1 and role='p' ;
```
Why are we using `mode='s'` here?

vbwagner replied:
Proibably it is not neccessary.  I've just tried to limit all possible variations so fault would be injected into just one segment.

vbwagner replied:
Removed this conditions from src/test/regress files. It doesn't affect test behavoir.

silent-observer commented on gpcontrib/gp_inject_fault/README:
```diff
@@ -133,16 +150,98 @@ status
 -----
 
-* A fault point applies to ALL callers of a function.  So special care
+ A fault point applies to ALL callers of a function. So special care
```
Still irrelevant changes present

vbwagner replied:
Fixed

Commit:
pgindent additions to faultinjector.c

Commit:
Remove declarations of Debug_dtm_action vars from guc.h

Commit:
Fix typo in the comments in gp_inject_fault--1.0.sql

Commit:
Add empty lines to make code more readable

Commit:
Fix indentation and typos in the faultinjector.h

Commit:
Remove some nonsignificant changes from gp_inject_fault README

Commit:
Improve tests and diffs readability

Improve formatting of calls to gp_inject_fault,
remove diffs with old expected results which are ignored
during computation of regression.diffs

RekGRpth commented on src/backend/tcop/postgres.c:
```diff
@@ -1652,7 +1576,7 @@ send_guc_to_QE(List *guc_list, bool is_restore)
 		{
 			/*
-			 * report error as warning 
+			 * report error as warning
```
remove unnecessary change

whitehawk commented on src/include/utils/guc.h:
```diff
@@ -400,14 +399,6 @@ typedef enum
 	DEBUG_DTM_ACTION_TARGET_LAST = 2
 }	DebugDtmActionTarget;
 
-extern char *Debug_dtm_action_sql_command_tag;
```
Can we remove data types `DebugDtmAction` and `DebugDtmActionTarget` as well?

vbwagner replied:
Yes they shoul be removed too.

vbwagner replied:
Fixed

silent-observer commented on src/test/regress/sql/distributed_transactions.sql:
```diff
@@ -13,7 +13,7 @@
 --
 --
--- We want to have an error between the point where all segments are prepared and our decision 
+-- We want to have an error between the point where all segments are prepared and our decision
```
More unnecessary changes here and on other test files too

vbwagner replied:
Fixed

Commit:
Remove debug_dtm_action-related types

Commit:
Improve formattig of sql code of isolation2 test

Commit:
Undo removal of end-of-line whitespace to reduce diff size

Commit:
Undo removal whitespace from end-of-line in postgres.c

RekGRpth commented on src/include/utils/faultinjector_lists.h:
```diff
@@ -65,6 +65,29 @@ FI_DDL_STATEMENT(CreateTablespaces, "create_tablespaces")
+FI_DDL_STATEMENT(MPPExecInsert, "MPPEXEC INSERT")
+FI_DDL_STATEMENT(MPPExecUpdate, "MPPEXEC UPDATE")
+FI_DDL_STATEMENT(MPPExecDelete, "MPPEXEC DELETE")
+FI_DDL_STATEMENT(Release,"RELEASE")
```
here and below add space after comma `,"` -> `, "`

vbwagner replied:
Fixed.

Commit:
Fix formatting of FI_DDL_STATEMENT macros

RekGRpth commented on src/backend/utils/misc/faultinjector.c:
```diff
@@ -281,12 +281,74 @@ checkBgProcessSkipFault(const char* faultName)
+
+	if (ddlStatement == DDLMax)
+		return FaultInjectorTypeNotSpecified;
+	return FaultInjector_InjectFaultIfSet_out_of_line(faultName, ddlStatement, "", "",
+													  nestingLevel);
+
```
```suggestion

	return FaultInjector_InjectFaultIfSet_out_of_line(faultName, ddlStatement, "", "",
													  nestingLevel);
```

RekGRpth commented on src/backend/utils/misc/faultinjector.c:
```diff
@@ -281,12 +281,74 @@ checkBgProcessSkipFault(const char* faultName)
+	DDLStatement_e val;
+};
+
+static const struct dtx2ddl dtxcmdmap[] = {
+	{DTX_PROTOCOL_COMMAND_NONE, DDLNotSpecified},
```
```suggestion
static const struct dtx2ddl dtxcmdmap[] =
{
	{ DTX_PROTOCOL_COMMAND_NONE, DDLNotSpecified },
```

vbwagner replied:
Ok

RekGRpth commented on src/backend/utils/misc/faultinjector.c:
```diff
@@ -281,12 +281,74 @@ checkBgProcessSkipFault(const char* faultName)
+
+FaultInjectorType_e
+FaultInjector_InjectFaultIfSet_out_of_line_DTX(
+											   const char *faultName,
```
```suggestion
FaultInjector_InjectFaultIfSet_out_of_line_DTX(const char *faultName,
```

vbwagner replied:
Ok

RekGRpth commented on src/backend/utils/misc/faultinjector.c:
```diff
@@ -281,12 +281,74 @@ checkBgProcessSkipFault(const char* faultName)
+	return FaultInjector_InjectFaultIfSet_out_of_line(faultName, ddlStatement, "", "",
+													  nestingLevel);
+
+}
```
```suggestion
		return FaultInjectorTypeNotSpecified;

	return FaultInjector_InjectFaultIfSet_out_of_line(faultName, ddlStatement, "", "",
													  nestingLevel);
}
```

vbwagner replied:
Don't understand what change suggested? Replace spaces to tabs or vice versa?

RekGRpth replied:
add new line after first return and remove one after second

vbwagner replied:
Yes, now I understand.

vbwagner replied:
Fixed

RekGRpth commented on src/include/utils/faultinjector.h:
```diff
@@ -99,19 +102,46 @@ extern FaultInjectorType_e FaultInjector_InjectFaultIfSet_out_of_line(
 extern char *InjectFault(
 	char *faultName, char *type, char *ddlStatement, char *databaseName,
-	char *tableName, int startOccurrence, int endOccurrence, int extraArg, int gpSessionid);
+	char *tableName, int startOccurrence, int endOccurrence, int extraArg, int gpSessionid, int nestingLevel);
```
may be format up to 80 lines?

vbwagner replied:
ok

RekGRpth commented on src/include/utils/faultinjector.h:
```diff
@@ -99,19 +102,46 @@ extern FaultInjectorType_e FaultInjector_InjectFaultIfSet_out_of_line(
 							   const char*				 databaseName,
-							   const char*				 tableName);
+							   const char*				 tableName,
+							   int						 nestingLevel);
+/* 
+ * Version of fault injection function which receives string tag
+ * instead of DDLStatement_e enumeration value.
+ */
```
may be move this comment and below one to c-file?

vbwagner replied:
I don't think so.  This comment documents prototype (usage) of public function. So it is better to keep it with declaration of function, not witth  it definition.

RekGRpth replied:
Throughout the code there are extensive comments before functions and almost no comments in the headers.

vbwagner replied:
In the faultinjector.h there are comments which make clear how these functions should be used.

I.e., line 96
```
/*
 * To check if a fault has been injected, use FaultInjector_InjectFaultIfSet().
 * It is designed to fall through as quickly as possible, when no faults are
 * activated.
 */
```

I think that my comments for functions *_SQL and *_DTX fell into same category.  May be they should be reformulated to make it look more like usage instruction, than function description

vbwagner replied:
Restructured comments moving old ones into C-file

Commit:
Fix minor formatting issues.

silent-observer commented on src/test/isolation2/expected/udf_exception_blocks_panic_scenarios.out:
```diff
@@ -64,29 +40,26 @@ CREATE FUNCTION
  p    | p              | 0       | s    | u      
- p    | p              | 1       | s    | u      
- p    | p              | 2       | s    | u      
+ m    | m              | 0       | s    | u      
```
Preferably these sorts of output changes should be avoided, since test success doesn't actually care about row ordering

vbwagner replied:
Fixed

silent-observer commented on src/test/regress/sql/udf_exception_blocks.sql:
```diff
@@ -425,7 +425,7 @@ CREATE VIEW test_invalid_schema_creation_tab_v1 AS SELECT f1 FROM test_invalid_s
 
 -- Tests truncate table success and failure case in function with exception
--- 
+--
```
Another unnecessary change

vbwagner replied:
Fixed

silent-observer commented on src/test/regress/sql/not_out_of_shmem_exit_slots.sql:
```diff
@@ -14,166 +14,200 @@
 -- s/session id \=\s*\d+/session id \= DUMMY/gm
 -- end_matchsubs
 
+
```
Another unnecessary change

vbwagner replied:
Fixed

Commit:
Minimize patch size by removing differences in out and sql

Remove erroneuously added empty line in
src/test/regress/sql/not_out_of_shmem_exit_slots.sql
and return back end-of-line space in
src/test/regress/sql/udf_exception_blocks.sql

Revert changes of line order in query output in
src/test/isolation2/expected/udf_exception_blocks_panic_scenarios.out
because this line order is ignored during diff.

Commit:
Restore end-of-line space in expected test result

Commit:
Move function description comments to c-file

Leave only brief usage instructions in faultinjector.h
Also reformat function FaultInjector_InjectFault_out_of_line_SQL.

whitehawk commented:
In the PR description:
'patch include' -> 'patch includes'
'form guc_gp.c' -> 'from guc_gp.c'

vbwagner commented:
> The PR description still requires adjustments according to our check-list.

Fixed paragraph formatting to 80 chars/per line. Any other suggestions?

vbwagner commented:
> In the PR description: 'patch include' -> 'patch includes' 'form guc_gp.c' -> 'from guc_gp.c'

Fixed
